Optimized :meth:`.Axes.coords_to_point`

chopan050 opened the pull request:
<!-- Thank you for contributing to Manim! Learn more about the process in our contributing guidelines: https://docs.manim.community/en/latest/contributing.html -->

## Overview: What does this pull request change?
<!-- If there is more information than the PR title that should be added to our release changelog, add it in the following changelog section. This is optional, but recommended for larger pull requests. -->
<!--changelog-start-->
As the title says, I optimized `Axes.coords_to_point`. Best results if combined with PR [#3285: Optimized :meth:`NumberLine.number_to_point`](https://github.com/ManimCommunity/manim/pull/3285).
<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?
<!-- Optional for bugfixes, small enhancements, and documentation-related PRs. Otherwise, please give a short reasoning for your changes. -->
- Mainly, I got rid of the `np.sum` and the list comprehension, and replaced them with a good ol' for loop, which is much more efficient in this case because there are only 2 or 3 coords to add.
- I avoided using `VMobject.__iter__` on `self.axes` by iterating directly over its `submobjects` attribute.
- Also, I removed one unnecessary addition and subtraction of `origin` by assigning first the result of `self.x_axis.number_to_point(coords[0])` by separate (which should not be a problem since `x_axis` is already being used explicitly to calculate `origin`.
- Finally, in general I worked to make the code more readable.

The changes I propose:

```py
    def coords_to_point(
        self, *coords: float | Sequence[float] | Sequence[Sequence[float]] | np.ndarray
    ) -> np.ndarray:
        coords = np.asarray(coords)
        origin = self.x_axis.number_to_point(
            self._origin_shift([self.x_axis.x_min, self.x_axis.x_max]),
        )

        # Is every component in coords in the format [xi, yi, zi]?
        # i.e. is coords in the format ([[x1 y1 z1] [x2 y2 z2] ...])? (True)
        #
        # Or is coords in the format (x, y, z) or ([x1 x2 ...], [y1 y2 ...], [z1 z2 ...])? (False)
        #
        # The latter is preferred.
        are_components_xyz = False

        # If coords is in the format ([[x1 y1 z1] [x2 y2 z2] ...]):
        if coords.ndim == 3:
            # Extract from original tuple: now the format is [[x1 y1 z1] [x2 y2 z2]]
            coords = coords[0]

            # If there's a single coord, extract it so that coords_to_point returns a single point
            if coords.shape[0] == 1:
                coords = coords[0]  # In this case, now coords = [x1 y1 z1]
            else:
                are_components_xyz = True
                # Transform coords into the format [[x1 x2 ...] [y1 y2 ...] [z1 z2 ...]]
                # for later processing.
                coords = coords.T

        # Now coords should be in the format [x y z], where each component is either a float or an ndarray
        points = self.x_axis.number_to_point(coords[0])
        other_axes = self.axes.submobjects[1:]
        for axis, nums in zip(other_axes, coords[1:]):
            points += axis.number_to_point(nums) - origin

        if are_components_xyz:
            return points
        return points.T  # Has no effect on a 1D array representing a single point
```

The results before and after the changes (using the same test scene I used for #3285). As you can see, I managed to cutoff around 0.70 seconds from `Axes.coords_to_point`. If the changes in the previous PR get accepted as well, there's even more speedup:

|Before|After|
|-|-|
![imagen](https://github.com/ManimCommunity/manim/assets/49853152/03e68c41-e5cc-4be1-ad6b-7d79926e383f) | ![imagen](https://github.com/ManimCommunity/manim/assets/49853152/222a69ba-4513-45a7-8106-db3dd5a0f032) |



## Links to added or changed documentation pages
<!-- Please add links to the affected documentation pages (edit the description after opening the PR). The link to the documentation for your PR is https://manimce--####.org.readthedocs.build/en/####/, where #### represents the PR number. -->


## Further Information and Comments
<!-- If applicable, put further comments for the reviewers here. -->



<!-- Thank you again for contributing! Do not modify the lines below, they are for reviewers. -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

behackl commented on manim/mobject/graphing/coordinate_systems.py:
```diff
@@ -1995,40 +1995,42 @@ def construct(self):
+        # Or is coords in the format (x, y, z) or ([x1 x2 ...], [y1 y2 ...], [z1 z2 ...])? (False)
+        #
+        # The latter is preferred.
+        are_components_xyz = False
```
I think this could be made even more explicit: `are_components_xyz` is not particularly self-explanatory. Why not `are_coordinates_transposed`?

chopan050 replied:
Yeah, I like that one better. The original `transposed` variable was not really clear IMO, but in hindsight `are_components_xyz` is not much better.

Gonna change that!

behackl reviewed (changes requested):
One minor suggestion, otherwise this is a fine improvement. Please take a look & thanks for your efforts!

chopan050 commented:
Done! I changed the variable name to `are_coordinates_transposed`, I liked that one better.

I also added some more commentaries and added an extra condition `points.ndim == 1` to the final return part. It bothered me a bit that the "single point" case falled into the `return points.T` situation and wanted to make more explicit that the single point should also be returned as is, but I forgot to add that before.

behackl reviewed (approved):
LGTM, thank you very much!
